bdk_cbf_example/src/cbf.rs

```diff
@@ -0,0 +1,166 @@
+use std::collections::HashSet;
+use std::{net, thread};
+
+use nakamoto::client::network::Services;
+use nakamoto::client::traits::Handle;
+use nakamoto::client::Handle as ClientHandle;
+use nakamoto::client::{chan, Client, Config, Domain, Event, Network};
+use nakamoto::net::poll;
+
+use bdk_cli::anyhow::{self, anyhow};
+
+use bdk_chain::bitcoin::{BlockHash, Script, Transaction};
+use bdk_chain::{BlockId, TxHeight};
+
+use crate::Domains;
+
+type Reactor = poll::Reactor<net::TcpStream>;
+
+// Things to handle:
+// - Block disconnected before filter processed, what to do?
+// - Rescan from 0
+
+pub struct CbfClient {
+    pub handle: ClientHandle<poll::reactor::Waker>,
+    pub setup: Option<CbfSetup>,
+}
+
+#[derive(Debug, Clone)]
+pub enum CbfEvent {
+    BlockMatched(BlockId, Vec<(Transaction, TxHeight)>),
+    BlockDisconnected(BlockId),
+    ChainSynced(Option<BlockId>),
+}
+
+pub struct CbfSetup {
+    pub events: chan::Receiver<Event>,
+    pub processed_height: Option<BlockId>,
+    pub blocks_matched: HashSet<BlockHash>,
+    pub peer_height: Option<u32>,
+}
+
+impl CbfClient {
+    pub fn new(network: Network, domains: Domains) -> anyhow::Result<Self> {
+        let domains = match domains {
+            Domains::IPv4 => vec![Domain::IPV4],
+            Domains::IPv6 => vec![Domain::IPV6],
+            Domains::Both => vec![Domain::IPV4, Domain::IPV6],
+        };
+
+        let config = Config {
+            network,
+            domains,
+            ..Config::default()
+        };
+
+        let client = Client::<Reactor>::new()?;
+        let handle = client.handle();
+
+        // Run the client on a different thread, to not block the main thread.
+        // Note that when we sync we rescan from the latest point of agreement,
+        // so it's not a problem if the client is running in the background
+        thread::spawn(|| client.run(config).unwrap());
+
+        Ok(CbfClient {
+            handle,
+            setup: None,
+        })
+    }
+
+    pub fn sync_setup(
+        &mut self,
+        scripts: impl Iterator<Item = Script>,
+        processed_height: Option<BlockId>,
```

Comment:
Instead of `processed_height` as an input, what do you think about using `local_chain: BTreeMap<u32, BlockHash>` and iterate in reverse until we find the last point of agreement?

Reply:
Nakamoto has `handle::Handle::get_block_by_height` so this should be doable.